File: 322_coin_change/coin_change.go
package main

import (
	"fmt"
	"math"
)

type Solution struct{}

func (s *Solution) coinChange(coins []int, amount int) int {
	dp := make([]int, amount+1)
	for i := range dp {
		dp[i] = math.MaxInt32
	}
	dp[0] = 0

	for i := 1; i <= amount; i++ {
		for _, coin := range coins {
			if i-coin >= 0 && dp[i-coin] != math.MaxInt32 {
				dp[i] = min(dp[i], 1+dp[i-coin])
			}
		}
	}

	if dp[amount] == math.MaxInt32 {
		return -1
	}
	return dp[amount]
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	s := Solution{}
	coins := []int{1, 2, 5}
	amount := 11
	fmt.Println(s.coinChange(coins, amount))
}

File: 322_coin_change/coin_change.py
from typing import List

class Solution:
    def coin_change(self, coins: List[int], amount: int) -> int:
        dp = [float("inf")] * (amount + 1)
        dp[0] = 0

        for i in range(1, amount + 1):
            for coin in coins:
                if i - coin >= 0:
                    dp[i] = min(dp[i], 1 + dp[i - coin])
        
        return dp[amount] if dp[amount] != float("inf") else -1


if __name__ == "__main__":
    solution = Solution()
    coins, amount = [1, 2, 5], 11
    print(solution.coin_change(coins, amount))
